packages/@ember/-internals/runtime/lib/system/object.js

```diff
@@ -44,21 +44,36 @@ Observable.apply(FrameworkObject.prototype);
 
 if (DEBUG) {
   let INIT_WAS_CALLED = symbol('INIT_WAS_CALLED');
+  let WILL_DESTROY_WAS_CALLED = symbol('WILL_DESTROY_WAS_CALLED');
   let ASSERT_INIT_WAS_CALLED = symbol('ASSERT_INIT_WAS_CALLED');
+  let ASSERT_WILL_DESTROY_WAS_CALLED = symbol('ASSERT_WILL_DESTROY_WAS_CALLED');
 
   FrameworkObject = class DebugFrameworkObject extends EmberObject {
     init() {
       super.init(...arguments);
       this[INIT_WAS_CALLED] = true;
     }
 
+    willDestroy() {
+      super.willDestroy(...arguments);
+      this[WILL_DESTROY_WAS_CALLED] = true;
+    }
+
     [ASSERT_INIT_WAS_CALLED]() {
       assert(
         `You must call \`super.init(...arguments);\` or \`this._super(...arguments)\` when overriding \`init\` on a framework object. Please update ${this} to call \`super.init(...arguments);\` from \`init\` when using native classes or \`this._super(...arguments)\` when using \`EmberObject.extend()\`.`,
         this[INIT_WAS_CALLED]
       );
     }
+
+    [ASSERT_WILL_DESTROY_WAS_CALLED]() {
+      assert(
+        `You must call \`super.willDestroy(...arguments);\` or \`this._super(...arguments)\` when overriding \`willDestroy\` on a framework object. Please update ${this} to call \`super.willDestroy(...arguments);\` from \`willDestroy\` when using native classes or \`this._super(...arguments)\` when using \`EmberObject.extend()\`.`,
+        this[WILL_DESTROY_WAS_CALLED]
+      );
+    }
```

Comment:
I worry that making this an assertion will be a breaking change, and I think we need to issue a deprecation for at least one LTS cycle first.

Reply:
I'm happy to look into any part of this which is actionable on my part if you let me know where to start.

Reply:
Rebased and took a crack at a deprecation here instead.